stingray/simulator/base.py

```diff
@@ -18,87 +28,152 @@ def simulate_times(lc, use_spline=False, bin_time=None):
-
-        lc_filt = counts[bin_start:bin_stop]
-        t_filt = times[bin_start:bin_stop]
-        length = t_filt[-1] - t_filt[0]
-
-        n_bin_filt = len(lc_filt)
-        n_to_simulate = n_bin_filt * np.max(lc_filt)
-        safety_factor = 10
+    Examples
+    --------
+    >>> t = [0.5, 1.5, 3.5]
+    >>> c = [100] * 3
+    >>> lc = Lightcurve(t, c, gti=[[0, 2], [3, 4]], dt=1.)
+    >>> times = simulate_times(lc, use_spline=True)
+    >>> np.all(np.diff(times) > 0)  # Output array is sorted
+    True
+    >>> np.all(times >= 0.)  # All times inside GTIs
+    True
+    >>> np.all(times <= 4.)
+    True
+    >>> np.any(times > 3.)
+    True
+    >>> np.any(times < 2.)
+    True
+    >>> np.any((times > 2.) & (times < 3.))  # No times outside GTIs
+    False
+    >>> lc.counts[0] = -3.
+    >>> simulate_times(lc)  # Test with one negative value in the lc
+    Traceback (most recent call last):
+        ...
+    ValueError: simulate_times can only work with...
+    """
+    return simulate_times_from_count_array(
+        lc.time, lc.counts, lc.gti, lc.dt, use_spline=use_spline)
 
-        if n_to_simulate > 10000:
-            safety_factor = 4.
 
-        n_to_simulate += safety_factor * np.sqrt(n_to_simulate)
-        n_to_simulate = int(np.ceil(n_to_simulate))
-        n_predict = ra.poisson(np.sum(lc_filt))
+def simulate_times_from_count_array(time, counts, gti, dt, use_spline=False):
+    """Simulate an event list, by using the inverse CDF method.
 
-        random_ts = ra.uniform(t_filt[0] - bin_time / 2,
-                               t_filt[-1] + bin_time / 2, n_to_simulate)
+    + Assume that the light curve is a probability density (must be positive
+      definite)
 
-        random_amps = ra.uniform(0, np.max(lc_filt), n_to_simulate)
+    + Calculate the CDF from the cumulative sum, and normalize it from 0 to 1
 
-        if use_spline:
-            lc_spl = sci.splrep(t_filt, lc_filt, s=np.longdouble(0), k=1)
-            pts = sci.splev(random_ts, lc_spl)
+    + Extract N random probability values from 0 to 1
 
-        else:
-            rough_bins = np.rint((random_ts - t0) / bin_time)
-            rough_bins = rough_bins.astype(int)
-            pts = lc_filt[rough_bins]
+    + Find the CDF values corresponding to these N values
 
-        good = random_amps < pts
-        len1 = len(random_ts)
-        random_ts = random_ts[good]
+    + Find the times corresponding to these N CDF values
 
-        len2 = len(random_ts)
-        random_ts = random_ts[:n_predict]
-        random_ts.sort()
+    Parameters
+    ----------
+    time: array-like
+    counts: array-like
+    gti: [[gti00, gti01], ..., [gtin0, gtin1]]
+    dt: float
 
-        new_nev = len(random_ts)
-        ev_list[nev:nev + new_nev] = random_ts[:]
-        nev += new_nev
-        bin_start += max_bin
+    Other Parameters
+    ----------------
+    use_spline : bool
+        Approximate the light curve with a spline to avoid binning effects
```

Comment:
I don't know what I think about the use of splines here. What's the default behaviour for that (that should probably be listed here)? Because the spline can help with binning effects, but you might just as well smooth out small-scale structure that then means the output PSD is no longer the same as the input PSD (because you've taken out some of the power at high frequencies.

Reply:
It's not smoothing, it's just interpolating differently between bins. It uses an order 3 polynomial instead of linear interpolation